nearcore/src/state_sync.rs

```diff
@@ -760,7 +760,7 @@ impl StateDumper {
 
     /// For each shard we're dumping state for, check whether the state sync header is already stored in the external storage,
     /// and set `header_to_dump` to None if so, so we don't waste time uploading it again.
-    async fn check_stored_headers(&mut self, dump: &mut DumpState) -> anyhow::Result<()> {
+    fn check_stored_headers(&self, dump: &mut DumpState) -> impl Future<Output = ()> {
```

Comment:
This change was required because after changing `&mut self` to `&self` in the signature the compiler started complaining that the future is not `Send`.

Reply:
This is fine, I guess, but when making a change like this it is always prudent to be aware that the change is *not* necessarily functionally equivalent with respect to the caller – previous code would execute the first statement in the function body the first time future is polled, whereas with a rewrite like this the initialization occurs at a function call time before the future is constructed. Both are valid things to need doing, so this cannot be applied blindly.

The only suggestion I have to better document the requirements/guarantees here is:

```suggestion
    fn check_stored_headers(&self, dump: &mut DumpState) -> impl Future<Output = ()> + Send {
```

Reply:
As alternative approach I would explore moving this method to `impl HeaderUploader { ... }` instead with the method taking `self: Arc<HeaderUploader>` as an argument. That would allow to keep using `async fn` which feels somewhat nicer still.